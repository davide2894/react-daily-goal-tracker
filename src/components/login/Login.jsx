import React, { useState } from "react";
import { loginWithEmailAndPassword } from "../../firebase";

function Login() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  function handleSubmit(evt) {
    evt.preventDefault();
    loginWithEmailAndPassword(email, password);
  }

  function handleEmailChange(evt) {
    setEmail(evt.target.value);
  }

  function handlePasswordChange(evt) {
    setPassword(evt.target.value);
  }

  return (
    <div
      data-testid="loginComponentTest"
      className="hi myAccount__form myAccount__form--login">
      <p>Login if you have already an account</p>
      <br></br>
      <form onSubmit={handleSubmit}>
        <div className="form-control">
          <label>Email</label>
          <input
            type="text"
            name="email"
            value={email}
            onChange={(evt) => handleEmailChange(evt)}
          />
        </div>
        <div className="form-control">
          <label>Password</label>
          <input
            type="password"
            name="password"
            value={password}
            onChange={(evt) => handlePasswordChange(evt)}
          />
        </div>
        <div className="form-control">
          <label></label>
          <button type="submit">Login</button>
        </div>
      </form>
    </div>
  );
}

export default Login;
