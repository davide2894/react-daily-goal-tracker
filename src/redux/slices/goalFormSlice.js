import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  show: false,
  goalTitle: "",
  goalScore: "",
};

export const goalFormSlice = createSlice({
  name: "goalFormSlice",
  initialState,
  reducers: {
    showForm: (state) => {
      state.show = true;
    },
    hideForm: (state) => {
      state.show = false;
    },
    setGoalTitle: (state, payload) => {
      state.goalTitle = payload;
    },
    setGoalScore: (state, payload) => {
      state.goalTitle = payload;
    },
  },
});

export const { showForm, hideForm, setGoalTitle, setGoalScore } =
  goalFormSlice.actions;
export default goalFormSlice.reducer;
