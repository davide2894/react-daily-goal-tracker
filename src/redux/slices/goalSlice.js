import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  goals: [
    {
      title: "my first goal",
      score: {
        max: 5,
        min: 0,
        actual: 0,
      },
      isComplete: false,
      id: "first",
    },
    {
      title: "my second goal",
      score: {
        max: 3,
        min: 0,
        actual: 1,
      },
      isComplete: false,
      id: "second",
    },
  ],
};

export const goalSlice = createSlice({
  name: "goalsSlice",
  initialState,
  reducers: {
    addGoal: (state, action) => {
      state.goals.push(action.payload);
    },
    incrementScore: (state, action) => {
      const goalToUpdate = state.goals.find(
        (goal) => goal.id === action.payload.id
      );

      if (goalToUpdate) {
        goalToUpdate.score.actual += 1;
      }
    },
    decrementScore: (state, action) => {
      const goalToUpdate = state.goals.find(
        (goal) => goal.id === action.payload.id
      );

      if (goalToUpdate) {
        goalToUpdate.score.actual -= 1;
      }
    },
  },
});

export const { addGoal, incrementScore, decrementScore } = goalSlice.actions;
export default goalSlice.reducer;
