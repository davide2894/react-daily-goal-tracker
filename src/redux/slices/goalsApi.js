// src/features/scores/scoresSlice.ts
import { createApi, fakeBaseQuery } from "@reduxjs/toolkit/query/react";
import {
  arrayUnion,
  collection,
  doc,
  getDoc,
  updateDoc,
  getDocs,
} from "firebase/firestore";
import { db } from "../../firebase";

export const firestoreApi = createApi({
  baseQuery: fakeBaseQuery(),
  tagTypes: ["Goals"],
  endpoints: (build) => ({
    fetchGoals: build.query({
      async queryFn(user) {
        try {
          let goals = [];
          const querySnapshot = await getDocs(
            collection(db, `/users/${user.userDocId}/user-goals/`)
          );
          querySnapshot.forEach((doc) => {
            goals.push(doc.data());
          });
          console.log({ data: goals });
          return { data: goals };
        } catch (error) {
          console.error(error.message);
          return { error: error.message };
        }
      },
      providesTags: ["Goals"],
    }),
  }),
});

export const { useFetchGoalsQuery } = firestoreApi;
