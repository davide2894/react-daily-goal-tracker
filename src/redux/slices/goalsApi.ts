// src/features/scores/scoresSlice.ts
import { createApi, fakeBaseQuery } from "@reduxjs/toolkit/query/react";
import {
  collection,
  doc,
  updateDoc,
  getDocs,
  setDoc,
  increment,
  deleteDoc,
  DocumentData,
} from "firebase/firestore";
import { db } from "../../firebase";

export const firestoreApi = createApi({
  baseQuery: fakeBaseQuery(),
  tagTypes: ["Goals"],
  endpoints: (builder) => ({
    fetchGoals: builder.query({
      async queryFn(user) {
        try {
          let goals: Array<DocumentData> = [];
          const querySnapshot = await getDocs(
            collection(db, `/users/${user.userDocId}/user-goals/`)
          );
          querySnapshot.forEach((doc) => {
            goals.push(doc.data());
          });
          console.log({ data: goals });
          return { data: goals };
        } catch (error) {
          console.error(error.message);
          return { error: error.message };
        }
      },
      providesTags: ["Goals"],
    }),
    // ts - ok
    addGoal: builder.mutation({
      async queryFn({ newGoal, currentUser }) {
        try {
          const docRef = doc(
            db,
            `/users/${currentUser.userDocId}/user-goals/`,
            newGoal.id
          );
          await setDoc(docRef, newGoal, { merge: true });
          return { data: newGoal };
        } catch (err) {
          console.log({ error: err });
          return { error: err };
        }
      },
      invalidatesTags: ["Goals"],
    }),
    // ts - ok
    decrementGoalScore: builder.mutation({
      async queryFn({ goalId, currentUser }) {
        try {
          const docRef = doc(
            db,
            `/users/${currentUser.userDocId}/user-goals/`,
            goalId
          );
          await updateDoc(docRef, {
            "score.actual": increment(-1),
          });
          return { data: "ok" };
        } catch (err) {
          return { error: err };
        }
      },
      invalidatesTags: ["Goals"],
    }),
    // ts - ok
    incrementGoalScore: builder.mutation({
      async queryFn({ goalId, max, min, currentUser }) {
        try {
          console.log({ goalId, currentUser });
          const docRef = doc(
            db,
            `/users/${currentUser.userDocId}/user-goals/`,
            goalId
          );
          await updateDoc(docRef, {
            "score.actual": increment(1),
          });
          const newActualScore = min + 1;
          if (newActualScore === max) {
            await updateDoc(docRef, {
              isComplete: true,
            });
          }
          return { data: "ok" };
        } catch (err) {
          return { error: err };
        }
      },
      invalidatesTags: ["Goals"],
    }),
    // ts - ok
    editGoal: builder.mutation({
      async queryFn({ updatedGoalData, currentUser }) {
        try {
          const docRef = doc(
            db,
            `/users/${currentUser.userDocId}/user-goals/`,
            updatedGoalData.id
          );
          await updateDoc(docRef, {
            "score.max": parseInt(updatedGoalData.newMaxScore),
            title: updatedGoalData.newTitle,
          });
          return { data: updatedGoalData };
        } catch (err) {
          return { error: err };
        }
      },
      invalidatesTags: ["Goals"],
    }),
    deleteGoal: builder.mutation({
      async queryFn({ goalId, currentUser }) {
        try {
          await deleteDoc(
            doc(db, `/users/${currentUser.userDocId}/user-goals/`, goalId)
          );
          return { data: "ok" };
        } catch (err) {
          return { error: err };
        }
      },
      invalidatesTags: ["Goals"],
    }),
    resetGoal: builder.mutation({
      async queryFn({ goal, currentUser }) {
        try {
          const docRef = doc(
            db,
            `/users/${currentUser.userDocId}/user-goals/`,
            goal.id
          );
          await updateDoc(docRef, {
            "score.actual": 0,
            isComplete: false,
          });
          return { data: goal };
        } catch (err) {
          return { error: err };
        }
      },
      invalidatesTags: ["Goals"],
    }),
  }),
});

export const {
  useFetchGoalsQuery,
  useAddGoalMutation,
  useEditGoalMutation,
  useDecrementGoalScoreMutation,
  useIncrementGoalScoreMutation,
  useDeleteGoalMutation,
  useResetGoalMutation,
} = firestoreApi;
