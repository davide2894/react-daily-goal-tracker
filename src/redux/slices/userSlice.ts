import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { LoginUserData, User } from "../../types";

export const userSlice = createSlice({
  name: "user",
  initialState: {
    user: {},
  } as User,
  reducers: {
    login: (state, action: PayloadAction<LoginUserData>) => {
      state.user = action.payload;
    },
    logout: (state) => {
      state.user = {};
    },
  },
});

export const { login, logout } = userSlice.actions;

// selectors
export const selectUser = (state) => state.user.user;
//   const posts = useSelector((state) => state.posts);

export default userSlice.reducer;
