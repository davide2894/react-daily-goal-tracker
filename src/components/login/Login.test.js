import { login } from "../../redux/slices/userSlice";
import Login from "./Login";
import { render, screen } from "@testing-library/react";

test("should render Login component", () => {
  render(<Login />);
  const loginComponentWrapper = screen.getByTestId("test");
  expect(loginComponentWrapper).toBeInTheDocument();
});
