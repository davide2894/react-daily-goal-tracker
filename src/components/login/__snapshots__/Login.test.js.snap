// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches its own snapshot 1`] = `
<div
  className="hi myAccount__form myAccount__form--login"
  data-testid="test"
>
  <p>
    Login if you have already an account
  </p>
  <br />
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-control"
    >
      <label>
        Email
      </label>
      <input
        name="email"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-control"
    >
      <label>
        Password
      </label>
      <input
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <div
      className="form-control"
    >
      <label />
      <button
        type="submit"
      >
        Login
      </button>
    </div>
  </form>
</div>
`;
